api/src/main/java/io/opencensus/metrics/DoubleGaugeMetric.java

```diff
@@ -0,0 +1,156 @@
+ * class YourClass {
+ *
+ *   private static final MetricRegistry metricRegistry = Metrics.getMetricRegistry();
+ *   DoubleGaugeMetric totalMemory = metricRegistry.addDoubleGaugeMetric(
+ *       "Total_memory", "Total CPU Memory", "1", new ArrayList<LabelKey>());
+ *
+ *   DataPoint defaultDataPoint = totalMemory.getDefaultDataPoint();
+ *
+ *   void doWork() {
+ *      defaultDataPoint.inc();
+ *      // Your code here.
+ *      defaultDataPoint.dec();
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Example 2: You can also use labels(keys and values) to track different types of metric.
+ *
+ * <pre>{@code
+ * class YourClass {
+ *
+ *   private static final MetricRegistry metricRegistry = Metrics.getMetricRegistry();
+ *
+ *   List<LabelKey> keys = Collections.singletonList(LabelKey.create("Type","desc"));
+ *   DoubleGaugeMetric totalMemory = metricRegistry.addDoubleGaugeMetric(
+ *       "Total_Memory", "Total CPU Memory", "1", keys);
+ *
+ *   List<LabelValue> usedMemory = Collections.singletonList(LabelValue.create("Used"));
+ *   DataPoint dataPoint = totalMemory.addDataPoint(usedMemory);
+ *
+ *   void doSomeWork() {
+ *      dataPoint.set(12.5);
+ *      // Your code here.
+ *      dataPoint.dec();
+ *   }
+ * }
+ * }</pre>
+ *
+ * @since 0.17
+ */
+public abstract class DoubleGaugeMetric {
+
+  /**
+   * Adds and returns new Data Point. This is more convenient form when you can want to manually
+   * increase and decrease values as per your service requirements. Must be passed the same number
+   * of label values are were passed to {@link MetricRegistry#addDoubleGaugeMetric} label keys.
+   *
+   * @param labelValues the list of label values.
+   * @return DataPoint the value of single gauge
+   * @since 0.17
+   */
+  public abstract DataPoint addDataPoint(List<LabelValue> labelValues);
+
+  /**
+   * Adds and returns new Data Point that reports the value of the object after the function. This
+   * is , slightly more common form of gauge is one that monitors some non-numeric object. The last
+   * argument establishes the function that is used to determine the value of the gauge when the
+   * gauge is collected. Must be passed the same number of label values are were passed to {@link
+   * MetricRegistry#addDoubleGaugeMetric} label keys.
+   *
+   * @param labelValues the list of label values.
+   * @param obj the state object from which the function derives a measurement.
+   * @param function the function to be called.
+   * @param <T> the type of the object upon which the function derives a measurement.
+   * @since 0.17
+   */
+  public abstract <T> void addDataPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function);
+
+  /**
+   * Returns a Data Point for a gauge without labels.
+   *
+   * @return DataPoint the value of default gauge
+   * @since 0.17
+   */
+  public abstract DataPoint getDefaultDataPoint();
+
+  /** The value of a single Gauge. */
+  public abstract static class DataPoint {
```

Comment:
Same here, maybe have one parent class `DataPoint` and two subclasses `DataPointDouble` and `DataPointLong`?